### DISCUSSION TITLE
Python build dependency question

### DISCUSSION THREAD

**Discussion by JakeSays** [ARMmbed/DAPLink#785] [General]
Other than running tests, what else is python use for when building daplink?

**Comment by flit** [comment]
On master, there are two supported build systems, mbed and progen. Both are Python based.

The Mbed build system is deprecated and will be dropped when the `experimental_compilers` branch is merged into master.

Progen is the [project-generator](/project-generator/project_generator) project. It uses yaml files to describe projects and define resuable components ([`projects.yaml`](/ARMmbed/DAPLink/tree/master/projects.yaml) and the [`records/`](/ARMmbed/DAPLink/tree/master/records) directory). Progen can generate IDE project files for Keil MDK, Eclipse Embedded CDT (and a few others that haven't been tested with DAPLink), or can generate GNU makefiles or CMake files. The primary reason we use it for DAPLink, compared to just using CMake for instance, is because it makes it very easy to compose the ~140 different firmware builds.

**Reply by JakeSays** [reply]
Ok that makes sense.

I also found a python tool called flashalgo (or something like that) - what does that do?

Basically I need to port daplink to a microchip ATSAME51 based board, but the process (not the code) seems unusually complicated. Am I missing something?

Actually I may be going about this the wrong way. All I need are basic SWD capabilities. What is the difference between this project and the CMSIS DAP code? Does this project build on top of that?

If so then I may just need that core cmsis code. I'm going to be using CAN instead of USB to communicate with the host.

**Reply by flit** [reply]
Using CAN instead of USB is really cool! The CMSIS-DAP protocol was designed to run over any transport, but I'm not aware of it ever being used with anything other than USB.

DAPLink does build on top of the reference CMSIS-DAP code (which is in `source/daplink/cmsis-dap/`). It adds virtual serial (USB CDC), optional mass storage flash programming, and WebUSB. And lots of preconfigured firmware builds. One way to think about the two is that the reference code is exactly that, reference code, while DAPLink builds production firmware.

Given that you only need SWD and are not using USB, going with the reference code would probably be more straightforward.

The `tools/flash_algo.py` script is part of some scripts to generate a C source file from the [.FLM style flash algorithms](https://arm-software.github.io/CMSIS_5/Pack/html/flashAlgorithm.html) used in CMSIS-Packs.

**Reply by JakeSays** [reply]
Thanks for the info! Greatly appreciated.
